jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java

```diff
@@ -223,13 +221,7 @@ public boolean exists()
                     if (in == null)
                     {
                         in = connection.getInputStream();
-                        Reference<InputStream> ref = new SoftReference<>(in);
-                        CLEANER.register(this, () ->
-                        {
-                            IO.close(ref.get());
-                            if (ON_SWEEP_LISTENER != null)
-                                ON_SWEEP_LISTENER.accept(ref);
-                        });
+                        CLEANER.register(this, new InputStreamCleaner(in));
```

Comment:
Rereading this code, I noticed the input stream gets closed if the `URLResource` gets garbage collected.

This means we get this:

```java
Resource webResource = resource.resolve("/web.xml");
assertTrue(webResource.exists());
InputStream inputStream = webResource.newInputStream();
webResource = null;

await().atMost(5, TimeUnit.SECONDS).until(() ->
{
    System.gc();
    return cleanedRefCount.get() > 0;
});

assertThat(IO.toString(inputStream), notNullValue()); //  this throws IOException: Stream closed
```

Reply:
Hmmm perhaps we need to weak reference the input stream itself? 

But it will generally be fine as I don't the reference to the input stream outlives the resource reference often.

Either way, this is better than what is committed.